### ISSUE TITLE
Problem with "propagate-constants" from given example.

### ISSUE THREAD

**Issue by alexgian** [mschuene/stratege#4]
Hi, some nice work here, potentially!
Excuse any ridiculously dumb questions, I'm new to this system.
However, this - following your example - does not give the right result.  Any ideas?

```
(ns stratego.core)
(require '[stratege.core :as s :refer [defrule rules]])

;; Second examples from github/mschuene/stratege main page

(def propagate-constants-rules
  (s/match-replace
   [:and true x] x
   [:and x true] x
   [:and false x] false
   [:and x false] false
   [:or true x] true
   [:or x true] true
   [:or false x] x
   [:or x false] x
   [:not false] true
   [:not true] false))
(def propagate-constants (s/downup propagate-constants-rules))
```
then
```
stratego.core=> (clojure.test/is (= true (propagate-constants [:and true [:or true false]])))

FAIL in () (form-init2163883587934145703.clj:1)
expected: (= true (propagate-constants [:and true [:or true false]]))
  actual: (not (= true nil))
false
```
Why does propagate-constants return nil?
Surely it should propagate to true.
Once again, apologies if I'm missing something too obvious.

**Comment by mschuene** [comment]
Hi, 
Yes, it should propagate to true. There seems to be something wrong with downup, I will investigate this. 

In the meantime, innermost works for this case (but is not as efficient) 

```
(def propagate-constants (innermost propagate-constants-rules))
(propagate-constants [:and true [:or true false]])
```

**Comment by alexgian** [comment]
Thanks!  Good to see the project is alive, could be a fantastic clojure tool.

However, even using innermost I'm getting some strange behaviour:
```
(def propagate-constants-rules
  (s/match-replace
   [:and true x] x
   [:and x true] x
   [:and false x] false
   [:and x false] false
   [:or true x] true
   [:or x true] true
   [:or false x] x
   [:or x false] x
   [:nor true x] false
   [:nor x true] false
   [:nor false x] [:not x]
   [:nor x false] [:not x]
   [:not false] true
   [:not true] false
   ))
(def propagate-constants (s/innermost propagate-constants-rules))
```
gives
```
$ lein repl
nREPL server started on port 42266 on host 127.0.0.1 - nrepl://127.0.0.1:42266
REPL-y 0.3.7, nREPL 0.2.12
Clojure 1.8.0
OpenJDK 64-Bit Server VM 1.8.0_111-8u111-b14-2ubuntu0.16.04.2-b14
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

stratego.core=> (propagate-constants [:and true true])
true
stratego.core=> (propagate-constants [:and true false])
[:and true false]
stratego.core=> (propagate-constants [:and false false])
[:and false false]
stratego.core=> (propagate-constants [:and false true])
[:and false true]
stratego.core=> (propagate-constants [:or true true])
true
stratego.core=> (propagate-constants [:or true false])
true
stratego.core=> (propagate-constants [:or false false])
[:or false false]
stratego.core=> (propagate-constants [:or false true])
true
stratego.core=> (propagate-constants [:nor true true])
[:nor true true]
stratego.core=> (propagate-constants [:nor true false])
[:nor true false]
stratego.core=> (propagate-constants [:nor false false])
true
stratego.core=> (propagate-constants [:nor false true])
[:nor false true]
stratego.core=> (propagate-constants [:not false])
true
stratego.core=> (propagate-constants [:not true])
[:not true]
stratego.core=>

stratego.core=> (propagate-constants [:and true [:or true false]])
true
stratego.core=> (propagate-constants [:or true [:or true false]])
true
stratego.core=> (propagate-constants [:or true [:and true false]])
true
stratego.core=> (propagate-constants [:and true [:and true false]])
[:and true false]
stratego.core=> (propagate-constants [:and false [:or true false]])
[:and false true]
stratego.core=> (propagate-constants [:or false [:or true false]])
true
stratego.core=> (propagate-constants [:or false [:and true false]])
[:and true false]
stratego.core=> (propagate-constants [:and false [:or true false]])
[:and false true]
stratego.core=>

stratego.core=> (propagate-constants [:and false [:and true false]])
[:and false [:and true false]]
```
Hope that sheds some light...

**Comment by alexgian** [comment]
Funnily enough, if you use the named rules system (first given example) everything seems to work fine, so far.

PS.  If you use s/downup rather than s/innermost with the _named_ rules, then they fail too, evaluating to nil as in the previous case.  So you're probably right about that.  

However, that seems to be a different problem to the rules failing to propagate at all, as demonstrated above.  (This does not seem to happen with named rules, only with the unnamed ones using s/match-replace. )

**Comment by mschuene** [comment]
Ok I found what was going on

First, the match-replace form should either be wrapped in (attempt ) for use with downup, because downup succeeds only when the strategy succeeded at all locations during the descend and ascend. In this case, once it gets to a leaf (the :or, true, or false) no left hand side of match-replace matches and thus fails, which also makes downup fail and return nil. You don't see this behavior with innermost, because innermost applies the rules until they no longer match and therefore already wraps the strategy with (attempt).

The bug you found as the use of a when-let instead of an (if (not (nil? ))) in the on-node function that is used for example to replace the current node with the instantiation of the right hand side of a matching rule. This led stratege to think that all rules with the right hand side false actually failed!

I pushed a new snapshot release to clojars, let me know if this resolves this issue.
